File: question/neetcode150/binary_search/search-in-rotated-sorted-array.py
from typing import List

class Solution:
    def search(self, a: List[int], t: int) -> int:
        f = a[0]

        if f <= t:
            i = 0
            n = len(a)

            while n:
                while i+n < len(a) and a[i+n] >= f and a[i+n] <= t:
                    i+=n
                n//=2
        else:
            n = len(a)-1
            i = n

            while n:
                while i-n >= 0 and a[i-n] < f and a[i-n] >= t:
                    i-=n
                n//=2

        return a[i] if a[i] == t else -1

File: question/neetcode150/binary_search/search-in-rotated-sorted-array.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int search(vector<int>& a, int t) {
        int n = a.size();
        int f = a[0];
        
        if (t>=f){
            int k = 0;

            for(int i=n;i;i/=2){
                while(k+i < n and a[k+i]>f and a[k+i]<=t){
                    k+=i;
                }
            }

            if(a[k] == t) return k;
        }else{
            int k = n-1;

            for(int i=n;i;i/=2){
                while(k-i >= 0 and a[k-i]<f and a[k-i]>=t){
                    k-=i;
                }
            }
            if(a[k] == t) return k;
        }

        return -1;
    }
};